idseq_dag/engine/pipeline_step.py

```diff
@@ -113,6 +116,54 @@ def wait_for_input_files(self):
                         time.sleep(5)
             self.input_files_local.append(flist)
 
+    def validate_input_files(self):
```

Comment:
Why the three methods, `validate_input_files`, `get_input_file_validation_errors` and `write_input_errors_json`? I think it would be simpler if you only had a single abstract method `validate_input_files` here which the pipeline_steps were responsible for implementing to produce the correct json. Also, that method should just add the json to `additional_files_to_upload` and the engine will take care of putting them in S3

Reply:
I wanted to share as much logic as possible and keep things consistent across pipeline steps. With my current approach, the PipelineStep only needs to return the validation errors. `validate_input_files` will output log statements, raise the RuntimeError, and package the error JSON consistently across all pipeline steps.

Also, I can't use `additional_files_to_upload` because that uploading only happens when the PipelineStep finishes. With this current approach, the PipelineStep still errors out prematurely.

I think this will look cleaner if I simplify to just an error type (see below).

Reply:
How about:
- PipelineStep has an attribute `errors` which is initialized as an empty dict
- PipelineStep has a method `validate_input_files` that checks if `self.errors` is empty or not. If not, it dumps `self.errors` and the step name into a json file, uploads the file to S3, and raises the runtime error
- children of PipelineStep override `validate_input_files` with a method that first adds any relevant entries to `self.errors` based on read counts and then calls `super()` which will ensure the upload. 
That way you have the shared logic for formatting and uploads, the individual steps only need to concern themselves with populating `self.errors`, and you avoid a spaghetti-like proliferation of methods :) I do think the current data structure and code structure are a bit overcomplicated

Reply:
I find the whole flow a bit confusing. 
Here is my proposal: 
 - Let's just put everything in one abstract method called validate_input_files
- This method will be called in thread_run method right after wait_for_input_files
- This method will throw an exception if the validation fails. We will define a CustomException for this purpose. 
- The pipeline_flow will capture this exception and upload an s3 error json file before it kills everything and exited. 

How does that sound?

Reply:
Hey @yunfangjuan, I originally tried that approach as it's the one we discussed when we synced. It turned out to be a little tricky because if you throw an exception in `thread_run`, that exception is being thrown inside a separate thread (exec_thread). Based on my understanding, it's fairly hard to catch an exception thrown inside a thread from the caller thread.

The Exception that `pipeline_flow.py` is catching at line 224 is actually being thrown inside `wait_until_all_done` at line 195, when the main thread realizes that the exec_thread is done but its status is not UPLOADED. It's not the same RuntimeException that's being thrown inside the Thread (this exception is never caught). 

The exec_thread can still communicate with the outside thread by setting self.status = INPUT_ERROR, so there are ways to do this if we want.

Is there a reason to introduce an extra CustomException, and have the pipeline_flow get involved as well? That seems to add additional steps to the flow. Happy to do it if it makes the abstraction cleaner.

Reply:
@cdebourcy That sounds good. I can do that.

Reply:
Hi @MarkAZhang 
I looked into the code a bit more. I think self.STATUS is the perfect place to place the input error. 

The following is the sample code and the status will be passed to wait_until_finished and throw a special type of exception for the PipelineFlow to handle. The reason for a special type of exception is to make PipelineFlow a central point of coordination.  We can even have pipeline flow to gather all kinds of different failures and uploads the failure analytics to S3 at the end.  


`
def thread_run(self):
        ''' Actually running the step '''
        self.status = StepStatus.STARTED
        self.wait_for_input_files()
        log.write("START STEP %s" % self.name)
         if self.validate_inputs() != GOOD_INPUT_RET_CODE:
             self.STATUS = BAD_INPUT
             return
        self.run()
        self.validate()
        self.save_progress()
        self.save_counts()
        log.write("FINISH STEP %s" % self.name)
        self.upload_thread = threading.Thread(target=self.uploading_results)
        self.upload_thread.start()
        self.status = StepStatus.FINISHED
`